File: JavaScript/310-Minimum Height Trees.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findMinHeightTrees = function(n, edges) {
    if (edges.length === 0){
        return [0];
    }
    
    let graph = new Map();
    for (let i=0; i<n; i++){
        graph[i] = new Set();
    }
    
    for (let [e1, e2] of edges){
        graph[e1].add(e2);
        graph[e2].add(e1);
    }

    let leaves = [];
    
    for (let i=0; i<n; i++){
        if (graph[i].size == 1){
            leaves.push(i);
        }
    }
    
    while (n>2){
        let leaveNum = leaves.length;
        n -= leaveNum
        for (let i=0; i<leaveNum; i++){
            let l = leaves.shift();
            for (let parent of graph[l]){
                graph[parent].delete(l);
                if (graph[parent].size === 1){
                    leaves.push(parent);
                }
            }
        }
    }
    return leaves;
};

File: C++/310-Minimum Height Trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if (edges.empty()) return {0};
        
        unordered_map<int, unordered_set<int>> graph;
        for (auto edge:edges) {
            int u = edge[0], v = edge[1];
            graph[u].insert(v);
            graph[v].insert(u);
        }
        
        queue<int> q;
        for (auto &kv:graph) {
            int k = kv.first;
            if (graph[k].size() == 1) q.push(k);
        }
        
        while (graph.size() > 2) {
            int length = q.size();
            while (length--) {
                int curr = q.front();
                q.pop();
                for (int nxt:graph[curr]) {
                    graph[nxt].erase(curr);
                    if (graph[nxt].size() == 1) {
                        q.push(nxt);
                    }
                }
                graph.erase(curr);
            }
        }
        
        vector<int> res;
        for (auto &kv:graph) {
            res.push_back(kv.first);
        }
        
        return res;
    }
};